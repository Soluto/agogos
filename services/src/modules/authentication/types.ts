import { ClientOptions } from 'jwks-rsa';

export interface IssuerConfig {
  authority: string;
  jwksConfig?: Partial<ClientOptions>;
  audience?: string | string[];
  authenticatedPaths: string[];
  description?: string;
}

export interface ApiKeyAuthConfig {
  header?: string;
  queryParam?: string;
  keys: Record<string, ApiKeyConfig>;
  authenticatedPaths: string[];
}

export interface ApiKeyConfig {
  name: string;
  description?: string;
  disabled?: boolean;
}

export interface AnonymousAuthConfig {
  publicPaths: string[];
  rejectAuthorizationHeader?: boolean;
}
export interface AuthenticationConfig {
  jwt?: Record<string, IssuerConfig>;
  apiKey?: ApiKeyAuthConfig;
  anonymous?: AnonymousAuthConfig;
}
