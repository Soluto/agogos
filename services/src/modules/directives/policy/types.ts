import {PolicyArgsObject, PolicyAttachments} from '../../resource-repository/types';

export type Policy = {
    namespace: string;
    name: string;
    args?: PolicyArgsObject;
};

// args here contain the final values after param injection
export type PolicyExecutionContext = {
    namespace: string;
    name: string;
    policyAttachments: PolicyAttachments;
    args?: PolicyArgsObject;
    query?: QueryResults;
};

export type QueryResults = {
    [name: string]: any;
};

export type PolicyExecutionResult = {
    done: boolean;
    allow?: boolean;
    query?: {
        type: string;
        code: string;
    };
};

export type GraphQLArguments = {
    [name: string]: any;
};
