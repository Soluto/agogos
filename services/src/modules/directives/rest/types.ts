export interface KeyValue {
  key: string;
  value: string;
}

export interface RestParams {
  url: string;
  method?: string;
  body?: string;
  bodyArg?: string;
  query?: KeyValue[];
  headers?: KeyValue[];
  timeoutMs?: number;
}
