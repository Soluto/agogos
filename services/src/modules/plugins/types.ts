import { BaseSchema } from '../base-schema';
import { ResourceGroup } from '../resource-repository';

export type ValueOrPromise<T> = T | Promise<T>;

export interface StitchPlugin {
  name: string;

  addArgumentInjectionGlobals?(): ValueOrPromise<Record<string, unknown>>;

  transformResourceGroup?(resourceGroup: ResourceGroup): ValueOrPromise<ResourceGroup>;

  transformBaseSchema?(baseSchema: BaseSchema): ValueOrPromise<BaseSchema>;
}
