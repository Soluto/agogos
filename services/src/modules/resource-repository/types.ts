import { RemoteSchema } from '../directives/gql';
import { Policy, LoadedPolicy } from '../directives/policy/types';

export type PolicyAttachments = Record<string, LoadedPolicy>;
export type PolicyQueryVariables = Record<string, unknown>;
export type PolicyArgsObject = Record<string, unknown>;

export interface ResourceGroup {
  schemas: Schema[];
  upstreams: Upstream[];
  upstreamClientCredentials: UpstreamClientCredentials[];
  policies: PolicyDefinition[];
  // policyAttachments are compiled from the Rego code in opa policies, they are not directly modified by users
  policyAttachments?: PolicyAttachments;
  basePolicy?: Policy;
  remoteSchemas?: RemoteSchema[];
}

export interface Resource {
  metadata: ResourceMetadata;
}

export interface ResourceMetadata {
  namespace: string;
  name: string;
}

export interface Schema extends Resource {
  schema: string;
}

export interface Upstream extends Resource {
  host: string;
  auth: {
    type: AuthType;
    activeDirectory: {
      authority: string;
      resource: string;
    };
  };
}

export interface UpstreamClientCredentials extends Resource {
  authType: AuthType;
  activeDirectory: {
    authority: string;
    clientId: string;
    clientSecret: string;
  };
}

export interface PolicyDefinition extends Resource {
  type: PolicyType;
  code: string;
  args?: PolicyArgsObject;
  query?: PolicyQuery;
  shouldOverrideBasePolicy?: boolean;
}

export interface PolicyQuery {
  gql: string;
  variables?: PolicyQueryVariables;
}

export interface FetchLatestResult {
  isNew: boolean;
  resourceGroup: ResourceGroup;
}

export type UpdateOptions = { registry: boolean };

export interface IResourceRepository {
  fetchLatest(): Promise<FetchLatestResult>;
  update(rg: ResourceGroup, options?: UpdateOptions): Promise<void>;
  writePolicyAttachment(filename: string, content: Buffer): Promise<void>;
}

enum AuthType {
  ActiveDirectory = 'ActiveDirectory',
}

export enum PolicyType {
  opa = 'opa',
}
