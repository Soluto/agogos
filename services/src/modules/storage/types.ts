export interface Storage {
  fileStats(filePath: string): Promise<FileStats>;
  writeFile(filePath: string, content: string | Buffer): Promise<void>;
  listFiles(folderPath: string): Promise<listFilesItem[]>;

  readFile(filePath: string): Promise<{ content: Buffer }>;
  readFile(filePath: string, options: { asString?: false }): Promise<{ content: Buffer }>;
  readFile(filePath: string, options: { asString?: true }): Promise<{ content: string }>;
  readFile(filePath: string, options?: { asString?: boolean }): Promise<{ content: Buffer | string }>;
}

export type FileStats = {
  lastModified: Date;
};

export type listFilesItem = {
  filePath: string;
  lastModified: Date;
};
