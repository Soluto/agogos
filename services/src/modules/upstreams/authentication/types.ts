import { Upstream, UpstreamClientCredentials } from '../../resource-repository';
import { ActiveDirectoryAuth } from '.';

export interface AuthenticationConfig {
  getUpstreamByHost: (host: string) => Upstream | undefined;
  getUpstreamClientCredentialsByAuthority: (authority: string) => UpstreamClientCredentials | undefined;
  activeDirectoryAuth: ActiveDirectoryAuth;
}
