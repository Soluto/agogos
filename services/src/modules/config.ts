import * as path from 'path';
import * as envVar from 'env-var';
import { AuthenticationConfig } from './authentication/types';

const envVarExt = envVar.from(process.env, {
  asSet: (value: string) => new Set(value.split(',')),
});

// General
export const httpPort = envVarExt.get('PORT').default('8080').asIntPositive();
export const logLevel = envVarExt.get('LOG_LEVEL').default('WARN').asString();
export const nodeEnv = envVarExt.get('NODE_ENV').default('development').asString();

// Repositories
export const useS3ResourceRepository = envVarExt.get('USE_S3_RESOURCE_REPOSITORY').default('false').asBoolStrict();
export const useFileSystemResourceRepository = envVarExt
  .get('USE_FS_RESOURCE_REPOSITORY')
  .default('true')
  .asBoolStrict();

// Resources
export const resourceUpdateInterval = envVarExt.get('RESOURCE_UPDATE_INTERVAL').default('60000').asIntPositive();

// GraphQL configuration
export const enableGraphQLTracing = envVarExt.get('GRAPHQL_TRACING').default('true').asBool();
export const enableGraphQLPlayground = envVarExt.get('GRAPHQL_PLAYGROUND').default('true').asBoolStrict();
export const enableGraphQLIntrospection = envVarExt.get('GRAPHQL_INTROSPECTION').default('true').asBoolStrict();

// Policies
export const tmpPoliciesDir = envVarExt
  .get('TMP_POLICIES_DIR')
  .default(path.resolve(process.cwd(), 'tmp/policies'))
  .asString();

export const ignorePolicies =
  envVar.get('NODE_ENV').required().asString() !== 'production' &&
  envVar.get('IGNORE_POLICIES').default('false').asBoolStrict();

// Authentication
const defaultAuthenticationConfig: AuthenticationConfig = {
  anonymous: {
    publicPaths: ['/metrics', '/.well-known/apollo/server-health', '/graphql'],
  },
};
export const authenticationConfig =
  (envVarExt.get('AUTHENTICATION_CONFIGURATION').asJsonObject() as AuthenticationConfig) ?? defaultAuthenticationConfig;

// Plugins
export const pluginsDir = envVar.get('PLUGINS_DIR').asString();
export const pluginsConfig = envVar.get('PLUGINS_CONFIGURATION').asJsonObject() as Record<string, unknown> | undefined;

// TODO:Check the list of apollo federation directives when upgrading apollo version
const defaultKnownApolloDirectives = 'key,extends,external,requires,provides';
export const knownApolloDirectives = envVarExt
  .get('KNOWN_APOLLO_DIRECTIVES')
  .default(defaultKnownApolloDirectives)
  .asSet();
