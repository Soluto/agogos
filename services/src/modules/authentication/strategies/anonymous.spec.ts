import { AnonymousAuthConfig } from '../types';
jest.mock('../../config');
import { authenticationConfig } from '../../config';
import anonymousAuthStrategy, { AnonymousAuthPartialRequest } from './anonymous';

interface TestCase {
  request: Partial<AnonymousAuthPartialRequest>;
  config?: AnonymousAuthConfig;
  shouldThrow?: boolean;
}

const url = '/some/path';

const testCases: [string, TestCase][] = [
  [
    'No anonymous auth config',
    {
      request: {},
      shouldThrow: true,
    },
  ],
  [
    'Unexpected path',
    {
      request: {
        raw: {
          url: '/unexpected/path',
        },
      },
      config: {
        publicPaths: [url],
      },
      shouldThrow: true,
    },
  ],
  [
    'Reject authorization header',
    {
      request: {
        headers: {
          authorization: 'Bearer JWT',
        },
        raw: {
          url,
        },
      },
      config: {
        rejectAuthorizationHeader: true,
        publicPaths: [url],
      },
      shouldThrow: true,
    },
  ],
  [
    'Valid request',
    {
      request: {
        raw: {
          url,
        },
      },
      config: {
        rejectAuthorizationHeader: true,
        publicPaths: [url],
      },
    },
  ],
  [
    'Valid request with authorization header',
    {
      request: {
        headers: {
          authorization: 'Bearer JWT',
        },
        raw: {
          url,
        },
      },
      config: {
        publicPaths: [url],
      },
    },
  ],
];

const defaultRequest: AnonymousAuthPartialRequest = {
  raw: {
    url: '/graphql',
  },
};

describe('Anonymous auth strategy', () => {
  test.each(testCases)('%s', async (_, { request, config, shouldThrow = false }) => {
    authenticationConfig.anonymous = config;
    try {
      await anonymousAuthStrategy({ ...defaultRequest, ...request });
      expect(shouldThrow).toBeFalsy();
    } catch (err) {
      expect(shouldThrow).toBeTruthy();
      expect(err).toMatchSnapshot();
    }
  });
});
