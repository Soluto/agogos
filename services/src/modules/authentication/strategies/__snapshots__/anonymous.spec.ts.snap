// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anonymous auth strategy No anonymous auth config 1`] = `[Error: Unauthorized]`;

exports[`Anonymous auth strategy Reject authorization header 1`] = `[Error: Unauthorized]`;

exports[`Anonymous auth strategy Unexpected path 1`] = `[Error: Unauthorized]`;
